File: day01/src/Part1.java
import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;



public class Part1
{
    public static final Pattern FIRST_NUMBER_PATTERN = Pattern.compile("^[^\\d]*(\\d)");
    public static final Pattern LAST_NUMBER_PATTERN = Pattern.compile("(\\d)[^\\d]*$");

    public static int line2number(String line) {
        Matcher fm = FIRST_NUMBER_PATTERN.matcher(line);
        if(!fm.find()) {
            System.err.format("No first digit match in string '%s'\n", line);
        }
        String firstDigit = fm.group(1);
        Matcher lm = LAST_NUMBER_PATTERN.matcher(line);
        if(!lm.find()) {
            System.err.format("No last digit match in string '%s'\n", line);
        }
        String lastDigit = lm.group(1);

        // System.out.format("    %s -> %s%s\n", line, firstDigit, lastDigit);

        return Integer.parseInt(firstDigit + lastDigit);
    }

    public static void main(String[] args) {
        Path filepath = FileSystems.getDefault().getPath(args[0]);

        try {
           int total = Files.lines(filepath)
                .filter(Predicate.not(String::isEmpty))
                .map(Part1::line2number)
                .reduce(0, Integer::sum);

            System.out.format("Part 1 Calibration: %d\n", total);
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }
}

File: day02/src/Part1.kt
package day02

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.util.function.Predicate;

import day02.game.Game



fun throwIsSubset(minThrow: Map<String, Int>, prompt: Map<String, Int>): Boolean {
    return prompt.entries.toList().all{ (color, amount) ->
        minThrow.getOrDefault(color, 0) <= amount
    }
}


val PROMPT = mapOf("red" to 12, "green" to 13, "blue" to 14)

fun main(args: Array<String>) {
    val filepath = FileSystems.getDefault().getPath(args[0]);

    try {
        val games = Files.lines(filepath)
            .filter(Predicate.not(String::isEmpty))
            .map(Game::parse)
            .filter{ game -> throwIsSubset(game.minimalThrow(), PROMPT) }
            .toList()
        
        val total = games.fold(0, {acc, game -> acc + game.id})

        System.out.format("Part 1 Total: %d\n", total);
    } catch (ioe: IOException) {
        ioe.printStackTrace();
    }
}